### PR TITLE
Exchange Integration

### PR DESCRIPTION
**Pull request by cracker0dks**
This will add Exchange Integration with all available calendar appointment features like -> add, change, delete and also getAvailability.

Note: I Added a .env var to define a default value (so not every user as to but is able to change):
```
# Used for the Exchange EWS integration
EXCHANGE_DEFAULT_EWS_URL=
```
For some reason this value is not pulled into the input field (it is set as default value at AddExchangeIntegration.tsx line 30 ). Maybe someone knows why and can fix this before merge?

The last commit (fix select event-types bug) is not really related to to exchange but will fix an issue with the radio buttons at /event-types (was not  able to select them).

Code is maybe not the cleanest so change and fix as you wish :)
Hope you like it.
Best regards!

PS: Fixed things so you get all your exchange calendars and be able to choose from them, also don't use appointments set to "free" at "show as"

### CODE REVIEW

File: public/static/locales/ru/common.json [hunk 1]
```diff
@@ -175,6 +175,7 @@
   "welcome_to_calcom": "Добро пожаловать в Cal.com",
   "welcome_instructions": "Скажите как вас зовут и в каком вы часовом поясе. Эту информацию потом можно будет изменить.",
   "connect_caldav": "Подключиться к серверу CalDav",
+  "connect_exchnage": "Подключиться к серверу Exchnage",
```
Comment (baileypumfleet) [inline public/static/locales/ru/common.json:178]:
```suggestion
  "connect_exchange": "Подключиться к серверу Exchnage",
```
Exchange is misspelled here :)

File: public/static/locales/es/common.json [hunk 2]
```diff
@@ -175,6 +175,7 @@
   "welcome_to_calcom": "Bienvenido a Cal.com",
   "welcome_instructions": "Cuéntanos cómo llamarte y dinos la zona horaria en la que estás. Podrás editarla más tarde.",
   "connect_caldav": "Conectar con el servidor de CalDav",
+  "connect_exchnage": "Conectar con el servidor de Exchange",
```
Comment (baileypumfleet) [inline public/static/locales/es/common.json:178]:
```suggestion
  "connect_exchange": "Conectar con el servidor de Exchange",
```
Exchange is misspelled here :)

File: yarn.lock [hunk 3]
```diff
@@ -5945,6 +5986,18 @@ modern-normalize@^1.1.0:
...
   resolved "https://registry.yarnpkg.com/modern-normalize/-/modern-normalize-1.1.0.tgz#da8e80140d9221426bd4f725c6e11283d34f90b7"
 
+moment-timezone@^0.5.26:
+  version "0.5.33"
+  resolved "https://registry.yarnpkg.com/moment-timezone/-/moment-timezone-0.5.33.tgz#b252fd6bb57f341c9b59a5ab61a8e51a73bbd22c"
+  integrity sha512-PTc2vcT8K9J5/9rDEPe5czSIKgLoGsH8UNpA4qZTVw0Vd/Uz19geE9abbIOQKaAQFcnQ3v5YEXrbSc5BpshH+w==
+  dependencies:
+    moment ">= 2.9.0"
+
+"moment@>= 2.9.0", moment@^2.24.0:
```
Comment (baileypumfleet) [inline yarn.lock:5996]: I'm not sure why moment is here, as it's not in the package.json dependencies. Ideally we'd like to stay away from moment, as we use day.js for everything, and moment is a very bloated dependency.

File: lib/integrations/getIntegrations.ts [hunk 4]
```diff
@@ -42,6 +42,14 @@ export const ALL_INTEGRATIONS = [
     description: "Video Conferencing",
     variant: "conferencing",
   },
+  {
+    installed: true,
+    type: "exchange_calendar",
+    title: "Exchange (EWS) Calendar",
+    imageSrc: "integrations/outlook.svg",
```
Comment (baileypumfleet) [inline lib/integrations/getIntegrations.ts:49]: Note to self: I'll change this to the actual Exchange logo

File: lib/integrations/Exchange/components/AddExchangeIntegration.tsx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      username: "",
+      password: "",
+    },
+  });
+  const [errorMessage, setErrorMessage] = useState("");
+  return (
+    <Dialog {...props}>
+      <DialogContent>
+        <DialogHeader
+          title="Connect to Exchange Server"
```
Comment (baileypumfleet) [inline lib/integrations/Exchange/components/AddExchangeIntegration.tsx:40]: These should ideally be extracted as translation keys. I'm happy to do this if you'd like

File: pages/api/integrations/exchange/add.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    // Check that user is authenticated
+    const session = await getSession({ req: req });
+
+    if (!session) {
+      res.status(401).json({ message: "You must be logged in to do this" });
+      return;
+    }
+
+    if (process.env.CALENDSO_ENCRYPTION_KEY == "") {
+      res.status(500).json({ message: "CALENDSO_ENCRYPTION_KEY must be definde in .env file!" });
```
Comment (baileypumfleet) [inline pages/api/integrations/exchange/add.ts:21]: Defined is misspelt here 😄

File: lib/integrations/Exchange/components/AddExchangeIntegration.tsx [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+} from "@components/Dialog";
+import { Form, TextField } from "@components/form/fields";
+import { Alert } from "@components/ui/Alert";
+import Button from "@components/ui/Button";
+
+type Props = {
+  onSubmit: () => void;
+};
+
+export const ADD_EXCHNAGE_INTEGRATION_FORM_TITLE = "addExchange";
```
Comment (baileypumfleet) [inline lib/integrations/Exchange/components/AddExchangeIntegration.tsx:20]:
```suggestion
export const ADD_EXCHANGE_INTEGRATION_FORM_TITLE = "addExchange";
```
Exchange is misspelt :)

File: lib/integrations/Exchange/components/AddExchangeIntegration.tsx [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+const AddExchangeIntegration = React.forwardRef<HTMLFormElement, Props>((props, ref) => {
+  const onSubmit = (event: any) => {
+    event.preventDefault();
+    event.stopPropagation();
+
+    props.onSubmit();
+  };
+
+  return (
+    <form id={ADD_EXCHNAGE_INTEGRATION_FORM_TITLE} ref={ref} onSubmit={onSubmit}>
```
Comment (baileypumfleet) [inline lib/integrations/Exchange/components/AddExchangeIntegration.tsx:121]: Exchange is misspelt :)

**Review by pumfleet — CHANGES_REQUESTED** [review]
Thank you so much for your contribution! This is a huge PR and we're super excited to get this added. I've added a few comments of some minor things I've found, and I'll wait for the rest of our team to review this. 😀

File: yarn.lock (hunk 3 above)
Comment (cracker0dks) [reply yarn.lock:5996]: I think this was added by the: ews-javascript-api module (For the DateTime.Parse function I guess)?

File: lib/integrations/Exchange/components/AddExchangeIntegration.tsx (hunk 5 above)
Comment (cracker0dks) [reply lib/integrations/Exchange/components/AddExchangeIntegration.tsx:40]: yeah sure 👍

File: lib/integrations/Exchange/components/AddExchangeIntegration.tsx (hunk 8 above)
Comment (cracker0dks) [reply lib/integrations/Exchange/components/AddExchangeIntegration.tsx:121]: Oh no, it was late :D

**Comment by PeerRich** [comment]
maybe @edanfesi can pick this up after he finished refactoring the integrations page 🙏

**Comment by vinyll** [comment]
Hey!
This is very promising and would be great for my company to move to an opensource booking solution.
Any plan to get this in a mergeable state?
Thanks so much for the work accomplished 🚀

File: lib/integrations/calendar/components/AddExchangeIntegration.tsx [hunk 9]
```diff
@@ -20,7 +20,8 @@ export function AddExchangeIntegrationModal(props: DialogProps) {
     defaultValues: {
       username: "",
       password: "",
-      url: process.env.EXCHANGE_DEFAULT_EWS_URL || "",
+      url:
+        process.env.EXCHANGE_DEFAULT_EWS_URL || "https://exchange.reutlingen-university.de/Ews/Exchange.asmx",
```
Comment (PeerRich) [inline lib/integrations/calendar/components/AddExchangeIntegration.tsx:24]: this should definitely not be here

Comment (cracker0dks) [reply lib/integrations/calendar/components/AddExchangeIntegration.tsx]: Sorry, pushed to the wrong repo, should be fixed now

File: lib/integrations/calendar/components/AddExchangeIntegration.tsx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  DialogContent,
+  DialogFooter,
+  DialogHeader,
+  DialogProps,
+} from "@components/Dialog";
+import { Form, TextField } from "@components/form/fields";
+import { Alert } from "@components/ui/Alert";
+import Button from "@components/ui/Button";
+
+export const ADD_APPLE_INTEGRATION_FORM_TITLE = "addExchangeIntegration";
```
Comment (zomars) [inline lib/integrations/calendar/components/AddExchangeIntegration.tsx:16]:
```suggestion
export const ADD_EXCHANGE_INTEGRATION_FORM_TITLE = "addExchangeIntegration";
```

**Comment by vinyll** [comment]
Wow, you guys are giving this PR a kick in the *** 🥇

**Comment by PeerRich** [comment]
sorry we are currently refactoring the app infrastructure

File: lib/integrations/getIntegrations.ts [hunk 11]
```diff
@@ -78,6 +79,14 @@ export const ALL_INTEGRATIONS = [
     description: "For personal and business calendars",
     variant: "calendar",
   },
+  {
+    installed: true,
```
Comment (zomars) [inline lib/integrations/getIntegrations.ts:83]: Do we need some env variables to consider this integration installed? Or does it work out of the box for self-hosted without any previous configuration?

Comment (cracker0dks) [reply lib/integrations/getIntegrations.ts:83]: yes, will work as long as the exchange server is running, so could be removed

Comment (PeerRich) [reply lib/integrations/getIntegrations.ts:83]: @cracker0dks

**Comment by PeerRich** [comment]
how do we continue with this PR @baileypumfleet @zomars @emrysal?

**Comment by aheil** [comment]
I am really looking forward for this PR! Thanks for this contribution!

**Comment by PeerRich** [comment]
hey everyone, we had some team changes in the last two weeks and are also working on an entire new app system which will make new integrations much much easier. @zomars will be working on it and then we'll look at how we can get this PR merged, don't you worry!


we're also hiring for an Apps and Integrations Engineer if anyone wants to get paid to work on Open Source 🎊  https://cal.com/jobs/full-stack-engineer-apps

**Comment by PeerRich** [comment]
@joeauyeung will re-open this in a new PR this week. We've made massive changes to the app store folder structure. Thank you for your contribution @cracker0dks, we may reuse some of the code you wrote and let you know once it's life!

File: yarn.lock (hunk 3 above)
Comment (gautamsi) [reply yarn.lock:6845]: I am open if someone want to contribute to 'ews-javascript-api' and replace moment (+ timezone) with something more lightweight like dayjs ore date-fns (or suggest the best alternative)

**Pull request closed without merging**